### PR TITLE
feat: Action Menu

### PR DESCRIPTION
**Pull request by jamil314**
[Country config pr](https://github.com/opencrvs/opencrvs-countryconfig/pull/286) for missing translations
[Farajaland pr](https://github.com/opencrvs/opencrvs-farajaland/pull/1135) with e2e changes
#7390

### CODE REVIEW

**Comment by Zangetsu101** [comment]
Can we create an API similar to this https://ui.shadcn.com/docs/components/dropdown-menu? It brings way more flexibility in it's usage. 
I'm pretty sure we can build our `ToggleMenu` component using the `dropdown-menu` quite easily which just shows that it's a better abstraction. 
@rikukissa @naftis What do you think? Are we willing to move towards these now? Using radix for example could improve our accessibility index quite a bit
Even still if we can just build a similar API as the `dropdown-menu,` it would still be a plus in my book

**Comment by naftis** [comment]
@Zangetsu101 Good idea, and yes please. We're already using something similar for example in `ListReview` and `Frame`:
```jsx
<ListReview>
	<ListReview.Header label="FIELD" value="INPUT" />
	<ListReview.Row label="Firstname(s)" value="Peter Jonathan" actions={LinkExample} />
	<ListReview.Row label="Lastname" value="Jones" actions={LinkExample} />
</ListReview>
  ```

File: packages/client/src/i18n/messages/views/action.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  view: MessageDescriptor
+  correctRecord: MessageDescriptor
+  archiveRecord: MessageDescriptor
+  reinstateRecord: MessageDescriptor
+  reviewCorrection: MessageDescriptor
+  reviewDeclaration: MessageDescriptor
+  updateDeclaration: MessageDescriptor
+  printDeclaration: MessageDescriptor
+  issueCertificate: MessageDescriptor
+}
```
Comment (naftis) [inline packages/client/src/i18n/messages/views/action.ts:26]: Do we need to define the types here actually or could we just directly pass the object as-is to `defineMessages`. Could help maintaining these

Comment (rikukissa) [reply packages/client/src/i18n/messages/views/action.ts:26]: Yea, I've occasionally removed these as they in reality make the typing worse :/

File: packages/client/src/views/RecordAudit/ActionMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+  )
+}
+
+const UnassignModal: React.FC<{
+  intl: IntlShape
+  close: (result: boolean | null) => void
+  assignment?: GQLAssignmentData
+  isDownloaded: boolean
+}> = ({ intl, close, assignment, isDownloaded }) =>
+  assignment && (
```
Comment (naftis) [inline packages/client/src/views/RecordAudit/ActionMenu.tsx:611]:
```suggestion

```

Let's have the calling end figure if the modal should be rendered to keep the component purer and move the conditional logic away from the view-components.

File: packages/client/src/views/RecordAudit/ActionMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+      ]}
+      show={true}
+      handleClose={() => close(null)}
+    >
+      {isDownloaded
+        ? intl.formatMessage(conflictsMessages.selfUnassignDesc)
+        : intl.formatMessage(conflictsMessages.regUnassignDesc, {
+            name: assignment.firstName + ' ' + assignment.lastName,
+            officeName: assignment.officeName
+          })}
```
Comment (naftis) [inline packages/client/src/views/RecordAudit/ActionMenu.tsx:644]: We could avoid having to pass an `assignment` object here by passing this component a `userName` and `officeName` for example. React cannot diff between objects so it helps with unnecessary re-renders.

File: packages/client/src/views/RecordAudit/ActionMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+  goToPage: typeof goToPage
+  goToPrintCertificate: typeof goToPrintCertificate
+  goToIssueCertificate: typeof goToIssueCertificate
+  goToUserProfile: typeof goToUserProfile
+}> = ({
+  declaration,
+  intl,
+  scope,
+  draft,
+  userDetails,
```
Comment (naftis) [inline packages/client/src/views/RecordAudit/ActionMenu.tsx:79]: Generally: let's not prop drill `intl` or other objects, as React is bad at diffing them and usually it makes purer and cleaner components by just passing the bare minimum to them what they need. If you need intl- here, use the `useIntl` hook.

I think this also applies to the goToFoo -functions, we can just import them directly without passing them as props and then dispatching them.

File: packages/components/src/Dropdown/Dropdown.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      display: 'flex',
+      justifyContent: 'center',
+      alignItems: 'center'
+    }}
+  >
+    <DropdownMenu>
+      <DropdownMenu.Trigger>
+        <PrimaryButton>Click Me</PrimaryButton>
+      </DropdownMenu.Trigger>
+      <DropdownMenu.Content {...args}>
```
Comment (naftis) [inline packages/components/src/Dropdown/Dropdown.stories.tsx:35]: Nice, the component API looks good and the component looks good as well!

File: packages/components/src/Dropdown/Dropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        return 'top: 50%; left: 100%; transform: translateY(-50%);'
+      case 'left':
+        return 'top: 50%; right: 100%; transform: translateY(-50%);'
+      default:
+        return ''
+    }
+  }}
+`
+
+const Label = styled.li`
```
Comment (naftis) [inline packages/components/src/Dropdown/Dropdown.tsx:73]: Good stuff using the correct semantic HTML elements 8)

File: packages/components/src/Dropdown/Dropdown.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            Dhaka
+          </DropdownMenu.Item>
+          <DropdownMenu.Item
+            onClick={() => alert('Helsinki is the capital of Finland')}
+          >
+            Helsinki
+          </DropdownMenu.Item>
+        </div>
+        <DropdownMenu.Separator weight={2} />
+        <div>
```
Comment (naftis) [inline packages/components/src/Dropdown/Dropdown.stories.tsx:51]: I think it's to group the items but the `div` seems a bit random. Should we use something named for it or is this a API choice that is done in other libraries as well?

File: packages/client/src/views/RecordAudit/ActionMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,646 @@
...
+  isDuplicate,
+  intl,
+  goToPage
+}) => {
+  const isPendingCorrection =
+    declarationStatus === EVENT_STATUS.CORRECTION_REQUESTED
+
+  const isReviewableDeclaration =
+    declarationStatus &&
+    [EVENT_STATUS.DECLARED, EVENT_STATUS.VALIDATED].includes(declarationStatus)
```
Comment (naftis) [inline packages/client/src/views/RecordAudit/ActionMenu.tsx:382]: @Zangetsu101 these `isFoo` logic stuff, could/should we move them to a centralized place near `useDeclaration`? I'm a bit worried that stuff like
```
  const isArchivable =
    declarationStatus &&
    [
      SUBMISSION_STATUS.IN_PROGRESS,
      SUBMISSION_STATUS.DECLARED,
      SUBMISSION_STATUS.VALIDATED,
      SUBMISSION_STATUS.REJECTED
    ].includes(declarationStatus as SUBMISSION_STATUS)
```
get defined in multiple places in the application and they start to mismatch and get outdated in the future.

The scope related stuff should lay in the `usePermissions` but that requires the scopes to be merged I think.

File: packages/client/src/components/interface/DownloadButton.tsx [hunk 9]
```diff
@@ -337,7 +337,12 @@ function DownloadButtonComponent(props: DownloadButtonProps & HOCProps) {
       <DownloadAction
         type="icon"
         id={`${id}-icon${isFailed ? `-failed` : ``}`}
-        onClick={onClickDownload}
+        onClick={
+          status === DOWNLOAD_STATUS.DOWNLOADED ||
+          (assignment && assignment.practitionerId !== practitionerId)
+            ? () => {}
+            : onClickDownload
+        }
```
Comment (naftis) [inline packages/client/src/components/interface/DownloadButton.tsx:345]: Maybe cleaner to do something like
```
const isDownloadable = ...

onClick={isDownloadable ? onDownloadClick : undefined}
```

File: packages/client/src/views/RecordAudit/ActionMenu.tsx [hunk 10]
```diff
@@ -0,0 +1,646 @@
...
+  )
+}
+
+const CorrectRecordAction: React.FC<
+  IActionItemCommonProps & IDeclarationProps
+> = ({ declarationId, declarationStatus, type, isDownloaded, intl, scope }) => {
+  const dispatch = useDispatch()
+
+  const isBirthOrDeathEvent =
+    type && [Event.Birth, Event.Death].includes(type.toLowerCase() as Event)
```
Comment (naftis) [inline packages/client/src/views/RecordAudit/ActionMenu.tsx:270]: > type && [Event.Birth, Event.Death].includes(type.toLowerCase() as Event)

Lets compare the actual event types together. If it means refactoring the calling end - we just need to spend time to do that. We have had issues with losing type safety when treating the event enum's as strings and lowercasing them

File: packages/client/src/views/RecordAudit/ActionMenu.tsx (hunk 8 above)
Comment (Zangetsu101) [reply packages/client/src/views/RecordAudit/ActionMenu.tsx:382]: We'll probably have to make some changes to this after the scope PR's merged in

File: packages/components/src/Dropdown/DropdownContext.tsx [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    } else if (e.key === 'ArrowUp') {
+      setFocusedIndex(getPreviousIndex)
+    }
+  }
+
+  useEffect(() => {
+    if (focusedIndex !== null && itemRefs.current[focusedIndex]) {
+      itemRefs.current[focusedIndex]?.focus()
+    }
+  }, [focusedIndex])
```
Comment (Zangetsu101) [inline packages/components/src/Dropdown/DropdownContext.tsx:86]: Does this need to be inside an `useEffect`? Maybe with an extra check like if the element doesn't have focus then bring focus to it, we can just call it on render?

File: packages/components/src/ToggleMenu/ToggleMenu.tsx
Comment (Zangetsu101) [inline packages/components/src/ToggleMenu/ToggleMenu.tsx:1]: Just looking at this file tells us that our new abstraction is just better :+1:

**Review by Zangetsu101 — CHANGES_REQUESTED** [review]

File: packages/components/src/Dropdown/Dropdown.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+      width: '100%',
+      height: '400px',
+      display: 'flex',
+      justifyContent: 'center',
+      alignItems: 'center'
+    }}
+  >
+    <DropdownMenu>
+      <DropdownMenu.Trigger>
+        <PrimaryButton>Click Me</PrimaryButton>
```
Comment (jpye-finch) [inline packages/components/src/Dropdown/Dropdown.stories.tsx:33]: PrimaryButton is deprecated. Please use Button

File: packages/components/src/Dropdown/Dropdown.stories.tsx (hunk 7 above)
Comment (jamil314) [reply packages/components/src/Dropdown/Dropdown.stories.tsx:51]: We don't actually have any logic for let's say: `<ItemGroup>`. We can always add it later if it is required. I'll remove the `<div>`s as they are not doing anything.

File: packages/components/src/Dropdown/Dropdown.tsx [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+  | 'span-all'
+  | 'end'
+
+const DropdownWrapper: React.FC<{ children: ReactNode; id: string }> = ({
+  children,
+  id
+}) => {
+  const { setDropdownName } = useDropdown()
+
+  setDropdownName(id)
```
Comment (Zangetsu101) [inline packages/components/src/Dropdown/Dropdown.tsx:128]: Let's directly pass this to `DropdownProvider` instead

**Review by Zangetsu101 — CHANGES_REQUESTED** [review]

**Review by Zangetsu101 — APPROVED** [review]

File: packages/client/src/views/OfficeHome/inProgress/inProgress.test.tsx [hunk 14]
```diff
@@ -797,7 +797,7 @@ describe('In Progress tab', () => {
           formatUrl(REVIEW_EVENT_PARENT_FORM_PAGE, {
             declarationId,
             pageId: 'review',
-            event: 'birth'
+            event: Event.Birth
```
[the remark this answers is not in the document]
Comment (jamil314) [reply packages/client/src/views/OfficeHome/inProgress/inProgress.test.tsx:800]: `Event` enum has `birth` in all lowercase

**Review by naftis — APPROVED** [review]

**Pull request merged**